File: packages/python/mythos_sdk/handshake.py
import os

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from jose import jwt
from jose.exceptions import ExpiredSignatureError, JOSEError, JWTError

from .jwks_cache import get_jwks, get_jwks_with_kid_fallback
from .logger import log_error
from .version import SDK_VERSION

_DEFAULT_API_URL = "https://api.mythos.work"
# Handshake tokens are validated by signature and purpose only.
_DECODE_OPTIONS = {"verify_aud": False, "verify_iss": False}


async def _validate_handshake_token(token: str) -> None:
    api_url = os.environ.get("MYTHOS_API_URL", _DEFAULT_API_URL)
    jwks = await get_jwks(api_url)
    try:
        payload = jwt.decode(token, jwks, algorithms=["ES256"], options=_DECODE_OPTIONS)
    except ExpiredSignatureError:
        raise
    except JWTError:
        jwks = await get_jwks_with_kid_fallback(api_url)
        payload = jwt.decode(token, jwks, algorithms=["ES256"], options=_DECODE_OPTIONS)

    if payload.get("purpose") != "handshake-check":
        raise JWTError("Token purpose is not handshake-check")


def create_handshake_router() -> APIRouter:
    router = APIRouter()

    @router.get("/.well-known/mythos-handshake")
    async def mythos_handshake(request: Request) -> JSONResponse:
        token = request.query_params.get("lt")
        if not token:
            return JSONResponse({"error": "Missing launch token"}, status_code=401)
        try:
            await _validate_handshake_token(token)
        except JOSEError:
            return JSONResponse({"error": "Invalid launch token"}, status_code=401)
        except Exception as err:
            log_error("handshake: unexpected error", err)
            return JSONResponse({"error": "Service unavailable"}, status_code=503)
        return JSONResponse({"ok": True, "sdk_version": SDK_VERSION})

    return router


handshake_router = create_handshake_router()

File: packages/node/src/handshake.ts
import { jwtVerify, errors } from 'jose';
import { Router } from 'express';
import type { Request, Response } from 'express';
import { getKeySet, getKeySetWithKidFallback } from './jwks-cache';

const SDK_VERSION = '0.1.0';
const DEFAULT_API_URL = 'https://api.mythos.work';

async function validateHandshakeToken(token: string): Promise<void> {
  const apiUrl = process.env.MYTHOS_API_URL ?? DEFAULT_API_URL;
  let keySet = await getKeySet(apiUrl);

  let payload;
  try {
    ({ payload } = await jwtVerify(token, keySet, { algorithms: ['RS256'] }));
  } catch (err: unknown) {
    if (!(err instanceof errors.JWKSNoMatchingKey)) throw err;
    keySet = await getKeySetWithKidFallback(apiUrl);
    ({ payload } = await jwtVerify(token, keySet, { algorithms: ['RS256'] }));
  }

  if (payload['purpose'] !== 'handshake-check') {
    throw new errors.JWTClaimValidationFailed(
      'Token purpose is not handshake-check',
      payload,
      'purpose',
      'check',
    );
  }
}

export function handshakeRoute(): Router {
  const router = Router();

  router.get('/.well-known/mythos-handshake', async (req: Request, res: Response) => {
    const raw = req.query['lt'];
    const first = Array.isArray(raw) ? raw[0] : raw;
    const token = typeof first === 'string' ? first : undefined;
    if (!token) {
      res.status(401).json({ error: 'Missing launch token' });
      return;
    }
    try {
      await validateHandshakeToken(token);
    } catch (err) {
      if (err instanceof errors.JOSEError) {
        res.status(401).json({ error: 'Invalid launch token' });
      } else {
        console.error('[mythos-sdk] handshake: unexpected error', err);
        res.status(503).json({ error: 'Service unavailable' });
      }
      return;
    }
    res.json({ ok: true, sdk_version: SDK_VERSION });
  });

  return router;
}
